web/api/2fa.go

```diff
@@ -169,6 +180,21 @@ func (ar *Router) FinalizeTFA() http.HandlerFunc {
 	}
 }
 
+func (ar *Router) verifyOTPCode(user model.User, otp string) (bool, error) {
+	result := false
+	if ar.tfaType == model.TFATypeApp {
+		totp := gotp.NewDefaultTOTP(user.TFAInfo().Secret)
+		result = totp.Verify(otp, int(time.Now().Unix()))
+	} else {
+		hotp := gotp.NewDefaultHOTP(user.TFAInfo().Secret)
+		result = hotp.Verify(otp, user.TFAInfo().HOTPCounter)
+		if user.TFAInfo().HOTPExpiredAt.Before(time.Now()) {
```

Comment:
I don't understand how `user.TFAInfo` works but I think it's better to check Expire time firstly, and then do other checks